### PR TITLE
feat(UniswapV3ExchangeAdapterV2): Add UniswapV3ExchangeAdapterV2 exchange adapter

### PR DESCRIPTION
**Pull request by 0xSachinK**
[From Slack] This adapter is necessary for the BasisTradingStrategyExtension and will be used there to acquire the spot asset. It supports multi-hop trades and allows user to specify `exactAmountOut` (as an alternative to `exactAmountIn`) 

- [x] Implement contract
- [x] Write unit tests
- [x] Write integration tests with TradeModule

### CODE REVIEW

File: test/protocol/integration/exchange/uniswapV3ExchangeAdapterV2.spec.ts [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+    let subjectFee1: BigNumberish;
+    let subjectToken2: Address;
+    let subjectFee2: BigNumberish;
+    let subjectToken3: Address;
+    let subjectFixIn: boolean;
+
+    beforeEach(async () => {
+      subjectToken1 = setup.wbtc.address;
+      subjectFee1 = 3000;
+      subjectToken2 = setup.weth.address;
```
Comment (cgewecke) [inline test/protocol/integration/exchange/uniswapV3ExchangeAdapterV2.spec.ts:232]: Maybe dai? 🐇

**Review by cgewecke — APPROVED** [review]
Really really nice!! Wow!

File: contracts/protocol/integration/exchange/UniswapV3ExchangeAdapterV2.sol [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+        data = abi.encodePacked(data, _path[_path.length - 1]);
+
+        // Encode fixIn
+        return abi.encodePacked(data, _fixIn);
+    }
+
+    /**
+     * Helper function to decode bytes to boolean. Similar to functions found in BytesLib.
+     */
+    function toBool(bytes memory _bytes, uint256 _start) public pure returns (bool) {
```
[the remark this answers is not in the document]
Comment (0xSachinK) [reply contracts/protocol/integration/exchange/UniswapV3ExchangeAdapterV2.sol:175]: Added a note in javadocs to make the reason more transparent.

**Pull request merged**